### DISCUSSION TITLE
Code structure

### DISCUSSION THREAD

**Discussion by laldoroty** [Roman-Supernova-PIT/diff-img#14] [General]
1. We start with a SN. This can (or must?) include: object ID, ra/dec, MJD range.
2. Input the RA/dec into function that pulls images that contain the RA/dec. (phrosty.utils.get_object_instances()) Save this table. 
3. Input MJD into function that splits images from step 2 into two tables: images that contain SN and those that don't. 
4. Calculate sky background model for all images in the two tables. 
5. Choose template. This is either single epoch or a coadd. Output is a single image. If coadd, make custom coadd for each science. Must subtract sky background using module from previous step before coadding. Cutout to 4088x4088 and aligned to the science image. 
6. Build template PSF. Retrieve PSFs for each image in the coadd at the SN location, rotate to science image, coadd them the same way as coadd. 
7. Sky subtraction for science image. 
8. Cross-convolve PSFs and images. 
9. SFFT subtraction step. 
10. Make decorrelation kernel. Inputs are the sigma_sky for the template and science images, and the two PSFs. 
11. Apply decorrelation kernel to the cross-convolved science image. 
12. If using stars to make ZPT, have to make the image for calculating the ZPT that has the same PSF as the difference image. So apply decorrelation kernel to the cross-convolved science image. 
13. PSF photometry on the stars in both the ZPT image and the difference image (this is a QA check--they should be 0). 
14. Make ZPT from the stars. S/N cut. 
15. Save ZPT. 
16. PSF photometry on the SN itself. 
17. Save catalog for that individual subtraction (filter, pointing, sca, mjd, mag, mag err, fit flux, flux err, zpt, ra, dec, x, y (fit and init))
18. Concatenate LC file from files from 17. 
19. QA & visualization: animations, LC plotting (truth vs. photometry recovery), analyze "stars" in different images, analyze SN in non-subtracted image

**Comment by laldoroty** [comment]
New structure for NN2 flux difference method (https://iopscience.iop.org/article/10.1086/491583/pdf) instead.

1. Start with SN OID, RA/Dec, MJD range
2. Get list of images that contain the RA/Dec
3. Input MJD into function that splits images from step 2 into two tables: images with the SN and images without.
4. Calculate sky background model for all images in the two tables.
5. Choose _N_ single-epoch templates. Rotate to science image.
6. Retrieve corresponding template PSFs. Rotate to science image. 
7. Sky subtraction for science image. 
8. Cross-convolve science PSF with each template PSF separately. 
9. SFFT subtraction _N_ times for each template image. 
10. Make decorrelation kernel for each subtracted image. 
11. Apply decorrelation kernels to corresponding subtracted images. 
12. Apply decorrelation kernels to corresponding cross-convolved science images (for ZPT calculation). 
13. PSF photometry on the stars in both the _N_ ZPT images and difference images (separately. photometry on stars in difference image should be 0 as a QA check). 
14. Make _N_ ZPTs from stars. 
15. Save ZPTs separately. 
16. PSF photometry on the SN itself in each difference image.
17. Save _N * (number of observations)_ catalogs for those subtractions (filter, pointing, sca, mjd, mag, mag err, fit flux, flux err, zpt, ra, dec, x, y (fit and init))
18. Concatenate step 17 by observation into _N_ catalogs. 
19. Build error matrix per eqns9-13 in NN2 paper.
20. Minimize chisq per eqn 2 in NN2 paper, A_{ij} is the final LC
21. Save A_{ij} to catalog with same columns as in step 17
22. QA and visualization: animations, LC plotting (truth vs photometry recovery, residuals), analyze "stars" in different images, analyze SN in non-subtracted image

**Reply by wmwv** [reply]
I suggestion making sure that N=1 is a supported, and perhaps even first, test case.

N=1 should work with single-image templates, and someday (but not yet) coadded templates.

**Reply by laldoroty** [reply]
Yes, definitely will test N=1 first.
